core/arch/instrlist.h

```diff
@@ -116,6 +121,22 @@ instrlist_get_translation_target(instrlist_t *ilist);
 void
 instrlist_set_our_mangling(instrlist_t *ilist, bool ours);
 
+DR_API
+/**
+ * All future instructions inserted into \p ilist will be predicated
+ * with \p pred. This is a convenience routine to make it easy to have
+ * emitted code from internal dr components predicated.
+ *
+ * \note only has an effect on ARM
+ */
+void
+instrlist_set_auto_predicate(instrlist_t *ilist, dr_pred_type_t pred);
```

Comment:
This sounds pretty good, though before that all the DR components (those methods similar to `dr_insert_clean_call()` which insert control flow or modify aflags, as well as whole modules like drreg) should be vetted such that they are known to work with it.

Also as a little test I globally enabled this flag in `drmgr` around the insertion event and ran some samples like `memtrace` and `instrace`, though they almost immediately crash. We need to look more into the interactions between `drreg` and auto predication before I feel comfortable enabling it globally.